idioms/mem-replace.md

```diff
@@ -0,0 +1,83 @@
+# `mem::replace` to keep owned values in changed enums
+
+## Description
+
+Say we have a `&mut MyEnum` which has (at least) two variants,
+`A { name: String, x: u8 }` and `B { name: String }`. Now we want to change
+`MyEnum::A` to a `B` if `x` is zero, while keeping `MyEnum::B` intact.
+
+We can do this without cloning the `name`.
+
+## Example
+
+```rust
+use std::mem;
+
+fn a_to_b(e: &mut MyEnum) {
+    // we mutably borrow `e` here. This precludes us from changing it directly
+    // as in `*e = ...`, because the borrow checker won't allow it. Of course
+    // we could just take a reference to `name` and clone that, but why pay an
+    // extra allocation for something we already have?
+    let have_name = match *e {
+        MyEnum::A(ref mut name, x) if x == 0 => {
+            // this takes out our `name` and put in an empty String instead
+            // note that empty strings don't allocate
+            Some(mem::replace(name, "".to_string()))
+        }
+        // nothing to do in all other cases
+        _ => None
+    };
+    // the mutable borrow ends here, so we can change `e`
+    if let Some(name) = have_name { *e = MyEnum::B { name } }
```

Comment:
`MyEnum::B { name: name }` at least until the [field init shorthand RFC](https://github.com/rust-lang/rfcs/pull/1682) stabilizes if accepted.